docs/design/datacontracts/data_descriptor.md

```diff
@@ -0,0 +1,310 @@
+
+## Types
+
+The types (both primitive types and structures described by structure descriptors) are classified as
+having either determinate or indeterminate type.  Determinate sizes may be used for pointer
+arithmetic.  Types with indeterminate size may not be.  Note that some sizes may be determinate, but
+/target specific/.  For example pointer types have a fixed size that varies by architecture.
+
+The following primitive types have indeterminate size: `string`.
+
+## Structure descriptors
+
+Each structure descriptor consists of:
+* a name
+* an optional size in bytes
+* a collection of field descriptors
+
+If the size is not given, the type has indeterminate size.  The size may also be given explicitly as
+"indeterminate" to emphasize that the type has indeterminate size.
+
+The collection of field descriptors may be empty.  In that case the type is opaque.  The primitive
+types may be thought of as opaque (for example: on ARM64 `nuint` is an opaque 8 byte type, `int64`
+is another opaque 8 byte type. `string` is an opaque type of indeterminate size).
+
+Type names must be globally unique within a single logical descriptor.
+
+### Field descriptors
+
+Each field descriptor consists of:
+* a name
+* an offset in bytes from the beginning of the struct
+* a type or the special type `variable`
+
+The name of a field descriptor must be unique within the definition of a structure.
+
+The offset may be negative.
+
+Two or more fields may have the same offets or imply that the underlying fields overlap.  The field
+offsets need not be aligned using any sort of target-specific alignment rules.
+
+Each field's type may refer to one of the primitive types or to any other type defined in the logical descriptor.
+
+If the field's type is `variable` it represents a byte-array of indeterminate size starting at the given offset.
+
+If a structure descriptor contains at least one field of indeterminate size, the whole structure
+must have indeterminate size.  Tooling is not required to, but may, signal a warning if a descriptor
+has a determinate size and contains indeterminate size fields.
+
+It is expected that tooling will signal a warning if a field specifies a type that does not appear
+in the logical descriptor.
+
+## Global value descriptors
+
+Each global value descriptor consists of:
+* a name
+* a type
+* a value
+
+The name of each global value must be unique within the logical descriptor.
+
+The type must be one of the determinate-size primitive types.
+
+The value must be a integral constant within the range of its type.  Signed values are two's
+complement.  Pointer values need not be aligned and need not point to adressable target memory.
+
+
+## Physical descriptors
+
+The phsyical descriptors are meant to describe /subsets/ of a logical descriptor and to compose.
+Each physical descriptor can name an ordered sequence of zero or more "baseline" descriptor which is then
+considered to comprise a piece of the overall logical descriptor.
```

Comment:
I'm fine with restricting to a single baseline.

As I was writing this, I was undecided on whether the baselines will be hand-coded or machine-generated.  If they're handcoded, I thought it would make sense to allow the json data descriptors to be arbitrarily composed (so for example you could have separate .json files for separate runtime subsystems that you blend together in the end to form a baseline for a particular release.  Each of the subsystems would depend on some "basic-types-64-bit.json" baseline that provided sizes/layouts for common types).

If the baselines are just machine-generated snapshots none of this is needed and we can just have a single baseline json file and a single binary blob to form a logical descriptor.